tests/unit/map_test.c

```diff
@@ -516,6 +516,154 @@ static void test_hashmap_degenerate_hash_fn(void)
+
+    // Test replacing
+    for (int i = 0; i < 3; ++i)
+    {
+        char* k = MallocStringKey('A' + i);
+        char* v = MallocStringKey('0' + i);
+        assert_int_equal(ArrayMapInsert(m, k, v), 1);
+    }  // size == 3
+
+    // Fill array up to size 14
+    for (int i = 0; i < 14 - 3; ++i)
+    {
+        ArrayMapInsert(m, MallocStringKey('a' + i), MallocStringValue('a' + i));
+    }
+
+    // Test (no) insertion for size > 14
+    for (int i = 0; i < 3; ++i)
+    {
+        char* k = MallocStringKey('\0');
+        char* v = MallocStringValue('\0');
+        assert_int_equal(ArrayMapInsert(m, k, v), 0);
+        free(k);
+        free(v);
+    }
+
+    ArrayMapDestroy(m);
+}
+
+void test_array_map_get(void)
+{
+    ArrayMap *m = ArrayMapNew(StringEqual_untyped, free, free);
+
+    // Fill with some test strings
+    for (int i = 0; i < 5; ++i)
+    {
+        char *k = MallocStringKey('A' + i);
+        char *v = MallocStringValue('A' + i);
+        assert_int_equal(ArrayMapInsert(m, k, v), 2);
+    }
+
+    // Get all the test strings
+    for (int i = 0; i < 5; ++i)
+    {
+        char k[] = ARRAY_STRING_KEY('A' + i);
+        MapKeyValue *pair = ArrayMapGet(m, k);
+        assert_true(pair != NULL);
+        assert_string_equal(pair->key, (char[]) ARRAY_STRING_KEY('A' + i));
+        assert_string_equal(pair->value, (char[]) ARRAY_STRING_VALUE('A' + i));
+    }
+
+    // Get non existent keys
+    for (int i = 0; i < 5; ++i)
+    {
+        char k[] = ARRAY_STRING_KEY('\0');
+        MapKeyValue *pair = ArrayMapGet(m, k);
+        assert_true(pair == NULL);
+    }
+
+    ArrayMapDestroy(m);
+}
+
+static void test_array_map_remove(void)
+{
+    ArrayMap *m = ArrayMapNew(StringEqual_untyped, free, free);
+
+    // Fill with some test strings
+    for (int i = 0; i < 5; ++i)
+    {
+        char *k = MallocStringKey('A' + i);
+        char *v = MallocStringValue('A' + i);
+        assert_int_equal(ArrayMapInsert(m, k, v), 2);
+    }
+
+    // Remove all keys one by one in reverse order
+    for (int i = 4; i >= 0; --i)
+    {
+        char k[] = ARRAY_STRING_KEY('A' + 4 - i);
+        assert_true(ArrayMapRemove(m, k));
+        assert_true(ArrayMapGet(m, k) == NULL);
+    }
+
+    // Try to remove non existent keys
+    for (int i = 0; i < 5; ++i)
+    {
+        char k[] = ARRAY_STRING_KEY('A' + i);
+        assert_false(ArrayMapRemove(m, k));
+    }
+
+    ArrayMapDestroy(m);
+}
+
+static void test_array_map_soft_destroy(void)
+{
+    ArrayMap *m = ArrayMapNew(StringEqual_untyped, free, free);
+
+    // Generate some pairs and keep track of values
+    char *values[5];
+
+    for (int i = 0; i < 5; ++i)
+    {
+        char *k = MallocStringKey('A' + i);
+        char *v = MallocStringValue('A' + i);
+        values[i] = v;
+        assert_int_equal(ArrayMapInsert(m, k, v), 2);
+    }
+
+    // Soft destroy and free them manually (should not double free)
+    // DEV: perhaps too... much?
+    ArrayMapSoftDestroy(m);
```

Comment:
Should something go wrong this will crash the map_test. While we will have definitive information of why soft-remove failed we will not see whether the following tests succeed/fail. So this unit_test acts like a possible blocker for the rest.

However we could move the array map to the end of the unit_test array with soft destroy at the very end so that even if soft destroy fails there we know nothing else follows.
Is this something of interest?

Reply:
Don't worry about this. If there's a memory access violation, we will have to fix it ASAP anyway.